src/CoreWCF.Primitives/src/CoreWCF/Channels/MtomMessageEncoder.cs

```diff
@@ -374,7 +377,13 @@ public override Message ReadMessage(ArraySegment<byte> buffer, BufferManager buf
 
             MtomBufferedMessageData messageData = _factory.TakeBufferedReader(this);
             messageData._contentType = contentType;
-            messageData.Open(buffer, bufferManager);
+
+            int bufferLength = (int)buffer.Length;
+            byte[] bytes = bufferManager.TakeBuffer(bufferLength);
+            buffer.CopyTo(bytes.AsSpan(0, bufferLength));
+
```

Comment:
We should avoid copying where necessary. Ideally MtomBufferedMessageData should handle a ReadOnlySequence natively

Reply:
@mconnew I agree. 

The issue i have is that `IBufferedMessageData` is a public type which exposes `Buffer` through a `ArraySegment<byte>`.  Would you be open to the following update of the IBufferedMessageData

```c#
public interface IBufferedMessageData
    {
        MessageEncoder MessageEncoder { get; }
        [Obsolete("Use ReadOnlyBuffer property", true)]
        ArraySegment<byte> Buffer { get; }
        ReadOnlySequence<byte> ReadOnlyBuffer { get; }
        XmlDictionaryReaderQuotas Quotas { get; }
        void Close();
        void EnableMultipleUsers();
        XmlDictionaryReader GetMessageReader();
        void Open();
        void ReturnMessageState(RecycledMessageState messageState);
        RecycledMessageState TakeMessageState();
    }
```
Introducing this property would allow to avoid this useless buffer allocation. 
Other buffer recopy you noticed during the review should also be solved and we should be able to drop `BufferManager` usages from the encoder implementation.
The missing piece now is a `Stream.Write(ReadOnlyMemory<byte> memory)`. i would be happy if you can link the code you have in NetNamedPipe implementation.

Reply:
@mconnew I push this [commit](https://github.com/CoreWCF/CoreWCF/commit/5106720621414b5d49047297eec45cc8a30aec1e) that implement the intention described above